lib/redirect-agent.js

```diff
@@ -0,0 +1,171 @@
+const { InvalidArgumentError } = require('./core/errors')
+const util = require('./core/util')
+const EventEmitter = require('events')
+const { getGlobalAgent } = require('./agent')
+const assert = require('assert')
+
+const kMaxRedirections = Symbol('maxRedirections')
+const kAgent = require('agent')
+
+class RedirectAgent extends EventEmitter {
+  constructor ({ maxRedirections = 10, agent = getGlobalAgent() } = {}) {
+    super()
+
+    if (maxRedirections != null && (maxRedirections < 0 || !Number.isInteger(maxRedirections))) {
+      throw new InvalidArgumentError('maxRedirections must be a positive number')
+    }
+
+    this[kMaxRedirections] = maxRedirections
+    this[kAgent] = agent
+
+    this[kAgent].on('connect', (...args) => this.emit(...args))
+    this[kAgent].on('disconnect', (...args) => this.emit(...args))
+  }
+
+  dispatch (origin, { maxRedirections = this[kMaxRedirections], ...opts }, handler) {
+    if (maxRedirections != null && (maxRedirections < 0 || !Number.isInteger(maxRedirections))) {
+      throw new InvalidArgumentError('maxRedirections must be a positive number')
+    }
+
+    if (!maxRedirections) {
+      return this[kAgent].dispatch(origin, opts, handler)
+    }
+
+    if (util.isStream(opts.body) && util.bodyLength(opts.body) === 0) {
```

Comment:
Shouldn't it be `!== 0`? Or am I missing something?

Reply:
If the stream is empty (i.e. ` === 0´) we just simply destroy and remove it. So it's correct. This way we can support `pipeline`.

Reply:
So in that case I would do a noop in `data` and ignore the `error`, as we don't probably care as well.

Reply:
Yea, neither of them should be ever invoke. I'm mostly guarding against broken streams in the callbacks.

Reply:
You're right. Better not to ignore broken stuff in the wild because they might cause leaks in users' software.

Reply:
In that case check what happens when the error is thrown. I think it should propagate up to the user level.

Reply:
I'm not sure we can propagate it to the user. In this case the best we can do is unhandled exception. Since it should never happen I think that's maybe fine. @mcollina wdyt?

Reply:
I can maybe fix this.